File: 动态规划/121. Best Time to Buy and Sell Stock/dp.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        /*
            DP：假设第i-1天的sum是能获得的累计收益，profit是第i天能获得的收益
            sum += profit，如果profit大于sum，则重新赋值sum，说明找到新的买点
            用maxprofit记录sum的最大值，不能返回sum，因为最后可能会减小
            时间复杂度：O(n)
            空间复杂度：O(n)
        */
        if(prices.size() < 2) return 0;
        int sum = 0, max_profit = 0;
        int profit;
        for(int i = 1; i < prices.size(); i++) {
            profit = prices[i] - prices[i-1];
            sum += profit;
            if(profit > sum) sum = profit;
            if(sum > max_profit) max_profit = sum;
        }
        return max_profit;
    }
};

File: 动态规划/[LeetCode 188 H] Best Time to Buy and Sell Stock IV/dp.js
/**
 * @param {number} k
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (k, prices) {
  const len = prices.length;
  const remain = k;
  const dp = [...Array(len + 1)].map(() =>
    [...Array(remain + 1)].map(() => [...Array(2)].fill(0))
  );

  for (let i = 0; i <= remain; i++) {
    dp[0][i][1] = i === remain - 1 ? -prices[0] : -10e5;
  }

  for (let i = 0; i < len; i++) {
    // no remain count
    for (let j = remain - 1; j >= 0; j--) {
      dp[i + 1][j][0] = Math.max(dp[i][j][0], dp[i][j][1] + prices[i]);
      dp[i + 1][j][1] = Math.max(dp[i][j + 1][0] - prices[i], dp[i][j][1]);
    }
  }

  let ans = 0;
  for (let i = 0; i <= remain; i++) {
    ans = Math.max(ans, dp[len][i][0], dp[len][i][1]);
  }
  return ans;
};
